File: LeetCode/LeetCode/Problems/5/Problem55.cs
using System;

public class Problem55
{
    public bool CanJump(int[] nums)
    {
        var consecutiveZeros = 0;
        for (int i = nums.Length - 2; i >= 0; i--)
            if (nums[i] == 0)
                consecutiveZeros++;
            else if (consecutiveZeros > 0)
                if (nums[i] >= consecutiveZeros + 1)
                    consecutiveZeros = 0;
                else
                    consecutiveZeros++;
        return consecutiveZeros == 0;
    }

    public bool CanJumpV2(int[] nums)
    {
        var reachable = 0;
        for (var i = 0; i < nums.Length; i++)
        {
            if (i > reachable)
                return false;
            reachable = Math.Max(reachable, i + nums[i]);
        }
        return true;
    }
}

File: leetcode_rust/src/problems/problem_55.rs
pub fn can_jump(nums: Vec<i32>) -> bool {
    let mut consecutive_zeros = 0;
    for i in (0..nums.len() - 1).rev() {
        if nums[i] == 0 {
            consecutive_zeros += 1;
        } else if consecutive_zeros > 0 {
            if nums[i] >= consecutive_zeros + 1 {
                consecutive_zeros = 0;
            } else {
                consecutive_zeros += 1;
            }
        }
    }
    consecutive_zeros == 0
}

pub fn can_jump_v2(nums: Vec<i32>) -> bool {
    let mut reachable = 0;
    for i in 0..nums.len() {
        if i > reachable {
            return false;
        }
        reachable = reachable.max(i + nums[i] as usize);
    }
    true
}
